File: Hw3/sol2.py
import numpy as np
class Solution():
    def randomized_partition(self,A,l,r):
        '''
        Find random pivot, and swap with it with A[l]. A[l] stores the pivot location during the subroutine
        '''
        random_ind=np.random.randint(l, high=r+1, size=1)[0]
        A[random_ind],A[l]=A[l],A[random_ind]
        p = A[l]
        i = l
        for j in range(l+1,r+1):
            if A[j]<=p:
                i=i+1
                A[i],A[j]=A[j],A[i]
        A[i],A[l]=A[l],A[i]
        return i
    def Rand_Select(self,A,l,r,i): # to find ith order statistic
        if l==r:
            return A[l]
        q=self.randomized_partition(A,l,r)
        k=q-l+1
        if k==i:
            return A[q]
        elif i<k:
            return self.Rand_Select(A,l,q-1,i)
        else:
            return self.Rand_Select(A,q+1,r,i-k)
    def partition(self,A,l,r,x):
        for m in range(l,r+1):
            if A[m]==x:
                break
        A[m],A[l]=A[l],A[m]
        p=A[l]
        i=l
        for j in range(l+1,r+1):
            if A[j]<=p:
                i=i+1
                A[i],A[j]=A[j],A[i]
        A[i],A[l]=A[l],A[i]
        return i

sol_obj = Solution()
A=[1,2,3,4,5,6,7,8,9,10]
i=7
print (sol_obj.Rand_Select(A,0,len(A)-1,i))

File: Hw3/sol2.cpp
#include <algorithm>
#include <iostream>
#include <random>
#include <vector>
#include <list>
#include <set>
#include <chrono>
#include <numeric>
using namespace std;
using namespace std::chrono;
void swap(int *p, int *q)
{
  // helper function
  int x = *p;
  *p = *q;
  *q = x;
}
int randomized_partition(int A[], int l, int r){
  // find random pivot
  srand(time(NULL));
  int random_ind = l + rand() % (r - l);
  // swap it with first  element
  swap(A[random_ind],A[l]);
  int p=A[l];
  int i=l;
  for (int j=l+1;j<r+1;j++){
    if (A[j]<=p){
      i=i+1;
      swap(A[i],A[j]);
    }
  }
  swap(A[i],A[l]);
  // return the pivot position
  return i;
}

int rand_select(int A[], int l, int r, int i){
  if (l==r){
    return A[l];
  }
  int q=randomized_partition(A,l,r);
  int k=q-l+1;
  if (k==i){
    return A[q];
  }
  else if (i<k){
    return rand_select(A,l,q-1,i);
  }else{
    return rand_select(A,q+1,r,i-k);
  }
}

int main()
{
  // Create a array with 1-100 numbers
  int A[101];
  for (int i=0; i<100; ++i){
    A[i]=i+1;
  }
  cout <<"Original permutation is ";
  for (int i=0; i<100; ++i){
    cout<<A[i]<<" ";
  }
  cout<<endl;
  random_device rd;
  mt19937 g(rd());
  // Shuffle the array with 1-100 numbers
  shuffle(A, A+100, g);
  cout <<"Shuffled permutation is ";
  for (int i=0; i<100; ++i){
    cout<<A[i]<<" ";
  }
  cout<<endl;
  int number_of_elements = 100;
  // test for different cases
  int order_statistic = 3;
  cout << order_statistic<< "rd order statistic is ";
  cout<<rand_select(A, 0, number_of_elements-1, order_statistic)<<endl;
  order_statistic = 20;
  cout << order_statistic<< "th order statistic is ";
  cout<<rand_select(A, 0, number_of_elements-1, order_statistic)<<endl;
  order_statistic = 42;
  cout << order_statistic<< "nd order statistic is ";
  cout<<rand_select(A, 0, number_of_elements-1, order_statistic)<<endl;
  order_statistic = 98;
  cout << order_statistic<< "th order statistic is ";
  cout<<rand_select(A, 0, number_of_elements-1, order_statistic)<<endl;
  order_statistic = 80;
  cout << order_statistic<< "th order statistic is ";
  cout<<rand_select(A, 0, number_of_elements-1, order_statistic)<<endl;
  return 0;
}
